hal_st/middlewares/ble_middleware/GattClientSt.cpp

```diff
@@ -143,6 +209,9 @@ namespace hal
 
         if (onDiscoveryCompletion)
             infra::Subject<services::GattClientDiscoveryObserver>::NotifyObservers(std::exchange(onDiscoveryCompletion, nullptr));
+
+        if (onDone)
```

Comment:
This has a reentrancy problem: If an observer is triggered via onDiscoveryCompletion, and that observer starts another action which sets onDone, that onDone gets executed immediately.
You should make a copy of onDone, and invoke that copy at this point.

Reply:
Fixed.